File: 798/C_Infected_Tree.py
import math,sys;input=sys.stdin.readline;from collections import Counter; inf= 1e18; mod =10**9+7; pi = math.pi; e = math.e; modinv = lambda x,m: pow(x,m-2,m); import bisect; import heapq; mod9 = 998244353

for _ in range(int(input().rstrip())):
    n=int(input().rstrip())
    adj={i:[] for i in range(1,n+1)}
    for i in range(n-1):
        u,v=map(int,input().rstrip().split())
        adj[u].append(v)
        adj[v].append(u)

    count = {i:0 for i in range(1,n+1)}
    vis = [False]*n
    def setc(i):
        c=0;vis[i-1]=True
        for j in adj[i]:
            if(not vis[j-1]):
                c+=setc(j)
        count[i]=c
        return c+1

    def ans(i):
        pos = []
        for j in adj[i]:
            if(count[j]<count[i]):
                pos.append(j)
        if(len(pos)==2):
            pos = []
            for j in adj[i]:
                if(count[j]<count[i]):
                    pos.append(j)
            return max(count[pos[0]]+ans(pos[1]),count[pos[1]]+ans(pos[0]))
        elif(len(pos)==1):
            return count[pos[0]]
        else:return 0
        
    setc(1)
    # print(count)
    print(ans(1))

File: 798/C_Infected_Tree.cpp
#include<bits/stdc++.h>

using namespace std;

//typedef long long int;
#define inf 1e18
#define mod 1000000007
#define pb push_back


int setc(int i, int* count, bool* vis, vector<int> *adj ){
    int c=0;
    vis[i-1]=true;
    for (int j : adj[i]){
        if(not vis[j-1])
            c+=setc(j, count,vis,adj);
    }
    count[i]=c;
    return c+1;
}

int ans(int i,int* count, vector<int> *adj ){
    vector<int> pos;
    for (int j : adj[i]){
        if(count[j]<count[i])
            pos.push_back(j);
    }
    if(pos.size()==2)
        return max(count[pos[0]]+ans(pos[1],count,adj),count[pos[1]]+ans(pos[0],count,adj));
    else if(pos.size()==1)
        return count[pos[0]];
    else return 0;
}

int main(int argc, char const *argv[])
{
    int T;
    cin>>T;
    
    for(int t=0; t<T; t++){
        int n;
        cin >>n;
        vector<int> adj[n+1];

        for(int i=0;i<n-1;i++){
            int u,v;
            cin >>u>>v;

            adj[u].push_back(v);
            adj[v].push_back(u);
        }

        int count[n+1];
        bool vis[n+1];
        for(int i=0;i<n+1;i++)vis[i]=false;
        setc(1,count,vis,adj);
        cout<<ans(1,count,adj)<<endl;
    }
    
    return 0;
}